query-service-impl/src/main/java/org/hypertrace/core/query/service/projection/ProjectionTransformation.java

```diff
@@ -315,6 +323,77 @@ private Filter rebuildFilterOmittingDefaults(
     return builder.clearChildFilter().addAllChildFilter(childFilters).build();
   }
 
+  private QueryRequest rebuildRequest(
+      QueryRequest original,
+      List<Expression> selections,
+      List<Expression> aggregations,
+      Filter filter,
+      List<Expression> groupBys,
+      List<OrderByExpression> orderBys) {
+
+    Set<AttributeExpression> attributeExpressionSet =
+        Stream.concat(
+                findComplexAttributeExpressionFromOrderBy(orderBys),
+                findComplexAttributeExpressionFromFilter(filter))
+            .collect(Collectors.toSet());
+
+    return rebuildRequestOmittingDefaults(
+        original,
+        selections,
+        aggregations,
+        createFilterForComplexAttributeExpression(attributeExpressionSet, filter),
+        groupBys,
+        orderBys);
+  }
+
+  private Stream<AttributeExpression> findComplexAttributeExpressionFromOrderBy(
+      List<OrderByExpression> orderByExpressionList) {
+    Stream.Builder<AttributeExpression> builder = Stream.builder();
+    for (OrderByExpression orderByExpression : orderByExpressionList) {
+      if (orderByExpression.getExpression().getValueCase() == ValueCase.ATTRIBUTE_EXPRESSION
+          && orderByExpression.getExpression().getAttributeExpression().hasSubpath()) {
+        builder.add(orderByExpression.getExpression().getAttributeExpression());
+      }
+    }
+    return builder.build();
+  }
+
+  private Stream<AttributeExpression> findComplexAttributeExpressionFromFilter(Filter filter) {
+    Stream.Builder<AttributeExpression> builder = Stream.builder();
```

Comment:
Here, it could look like:

```java
if (QueryRequestUtil.isMapSubpathAttributeExpression(filter.getLhs())) {
  return Set.of(filter.getLhs());
}

return filter.getChildFilterList().stream()
  .map(this::findComplexAttributeExpressionFromFilter)
  .flatMap(Collection::stream)
  .collect(Collectors.toUnmodifiableSet);
}
```

(It's actually relatively rare that you actually need to construct a stream by hand - we almost always are transforming existing data)

Reply:
resolved